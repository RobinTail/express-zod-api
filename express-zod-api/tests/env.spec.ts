import createHttpError from "http-errors";
import * as R from "ramda";
import { z } from "zod";

describe("Environment checks", () => {
  describe("Zod Dates", () => {
    test.each(["2021-01-32", "22/01/2022", "2021-01-31T25:00:00.000Z"])(
      "should detect invalid date %#",
      (str) => {
        expect(z.date().safeParse(new Date(str)).success).toBeFalsy();
        expect(z.string().date().safeParse(str).success).toBeFalsy();
        expect(z.string().datetime().safeParse(str).success).toBeFalsy();
        expect(z.iso.date().safeParse(str).success).toBeFalsy();
        expect(z.iso.datetime().safeParse(str).success).toBeFalsy();
      },
    );
  });

  describe("Zod checks/refinements", () => {
    test.each([
      z.string().email(),
      z.email(),
      z.number().int(),
      z.int(),
      z.int32(),
      z.int().max(1000),
    ])("Snapshot control $constructor.name definition", (schema) => {
      const snapshot = R.omit(["version"], schema._zod);
      expect(snapshot).toMatchSnapshot();
    });
  });

  describe("Zod imperfections", () => {
    test("discriminated unions are not depicted well", () => {
      expect(
        z.toJSONSchema(
          z.discriminatedUnion("status", [
            z.object({ status: z.literal("success"), data: z.any() }),
            z.object({
              status: z.literal("error"),
              error: z.object({ message: z.string() }),
            }),
          ]),
        ),
      ).not.toHaveProperty("discriminator");
    });

    test("bigint is not representable", () => {
      const json = z.toJSONSchema(z.bigint(), { unrepresentable: "any" });
      expect(R.omit(["$schema"], json)).toEqual({});
    });

    test("circular object schema has a getter since 4.0.15", () => {
      const schema = z.object({
        name: z.string(),
        get features() {
          return schema.array();
        },
      });
      expect(
        Object.getOwnPropertyDescriptors(schema._zod.def.shape).features,
      ).toHaveProperty("get", expect.any(Function));
    });

    test("ZodError inequality", () => {
      const issue: z.core.$ZodIssue = {
        code: "invalid_type",
        expected: "string",
        input: 123,
        path: [],
        message: "expected string, received number",
      };
      const error = new z.ZodError([issue]);
      const real = new z.ZodRealError([issue]);
      expect(error).not.toBeInstanceOf(Error); // and this is important
      expect(real).toBeInstanceOf(Error);
      expect(real).toBeInstanceOf(z.ZodError); // important inheritance
      expect(error).toHaveProperty("message");
      expect(real).toHaveProperty("message");
    });
  });

  describe("Zod new features", () => {
    test("ZodError equality", () => {
      try {
        z.number().parse("test");
      } catch (caught) {
        const returned = z.number().safeParse("test").error;
        expect(returned).toEqual(caught);
        expect(returned).toBeInstanceOf(z.ZodError);
        expect(caught).toBeInstanceOf(z.ZodError);
        expect(returned).toBeInstanceOf(Error);
        expect(caught).toBeInstanceOf(Error);
      }
    });

    test("meta() merge, not just overrides", () => {
      const schema = z
        .string()
        .meta({ examples: ["test"] })
        .describe("some")
        .meta({ title: "last" });
      expect(schema.meta()).toMatchSnapshot();
    });

    test("object shape conveys the keys optionality", () => {
      const schema = z.object({
        one: z.boolean(),
        two: z.boolean().optional(),
        three: z.boolean().default(true),
        four: z
          .boolean()
          .optional()
          .transform(() => false),
      });
      expect(Object.keys(schema._zod.def.shape)).toEqual([
        "one",
        "two",
        "three",
        "four",
      ]);
      expect(schema._zod.def.shape.one._zod.optin).toBeUndefined();
      expect(schema._zod.def.shape.one._zod.optout).toBeUndefined();
      expect(schema._zod.def.shape.two._zod.optin).toBe("optional");
      expect(schema._zod.def.shape.two._zod.optout).toBe("optional");
      expect(schema._zod.def.shape.three._zod.optin).toBe("optional");
      expect(schema._zod.def.shape.three._zod.optout).toBe(undefined);
      expect(schema._zod.def.shape.four._zod.optin).toBe("optional");
      expect(schema._zod.def.shape.four._zod.optout).toBe(undefined);
      expectTypeOf<z.input<typeof schema>>().toEqualTypeOf<{
        one: boolean;
        two?: boolean | undefined;
        three?: boolean | undefined;
        four?: boolean | undefined;
      }>();
      expectTypeOf<z.output<typeof schema>>().toEqualTypeOf<{
        one: boolean;
        two?: boolean | undefined;
        three: boolean;
        four: boolean;
      }>();
    });

    test("coerce is safe for nullable and optional", () => {
      const boolSchema = z.coerce.boolean();
      expect(boolSchema.isOptional()).toBeTruthy();
      expect(boolSchema.isNullable()).toBeTruthy();
    });

    /** @link https://github.com/colinhacks/zod/issues/4274 */
    test.each(["input", "output"] as const)(
      "%s examples of transformations",
      (io) => {
        const schema = z
          .string()
          .meta({ examples: ["test"] })
          .transform(Number)
          .meta({ examples: [4] });
        expect(
          z.toJSONSchema(schema, { io, unrepresentable: "any" }),
        ).toMatchSnapshot();
      },
    );

    test("meta id goes directly to depiction", () => {
      expect(z.toJSONSchema(z.string().meta({ id: "uniq" }))).toMatchSnapshot();
    });
  });

  describe("Vitest error comparison", () => {
    test("should distinguish error instances of different classes", () => {
      expect(createHttpError(500, "some message")).not.toEqual(
        new Error("some message"),
      );
    });

    test("should distinguish HTTP errors by status code and message", () => {
      expect(createHttpError(400, "test")).not.toEqual(
        createHttpError(500, "test"),
      );
      expect(createHttpError(400, "one")).not.toEqual(
        createHttpError(400, "two"),
      );
      expect(createHttpError(400, new Error("one"))).not.toEqual(
        createHttpError(400, new Error("two")),
      );
    });

    test("should distinguish error causes", () => {
      expect(new Error("test", { cause: "one" })).not.toEqual(
        new Error("test", { cause: "two" }),
      );
      expect(
        createHttpError(400, new Error("test", { cause: "one" })),
      ).not.toEqual(createHttpError(400, new Error("test", { cause: "two" })));
    });
  });
});
