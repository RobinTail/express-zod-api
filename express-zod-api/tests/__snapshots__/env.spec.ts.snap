// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Environment checks > Zod checks/refinements > Snapshot control 'ZodEmail' definition 1`] = `
{
  "bag": {
    "format": "email",
    "pattern": /\\^\\(\\?!\\\\\\.\\)\\(\\?!\\.\\*\\\\\\.\\\\\\.\\)\\(\\[A-Za-z0-9_'\\+\\\\-\\\\\\.\\]\\*\\)\\[A-Za-z0-9_\\+-\\]@\\(\\[A-Za-z0-9\\]\\[A-Za-z0-9\\\\-\\]\\*\\\\\\.\\)\\+\\[A-Za-z\\]\\{2,\\}\\$/,
  },
  "check": [Function],
  "constr": [Function],
  "def": {
    "abort": false,
    "check": "string_format",
    "format": "email",
    "pattern": /\\^\\(\\?!\\\\\\.\\)\\(\\?!\\.\\*\\\\\\.\\\\\\.\\)\\(\\[A-Za-z0-9_'\\+\\\\-\\\\\\.\\]\\*\\)\\[A-Za-z0-9_\\+-\\]@\\(\\[A-Za-z0-9\\]\\[A-Za-z0-9\\\\-\\]\\*\\\\\\.\\)\\+\\[A-Za-z\\]\\{2,\\}\\$/,
    "type": "string",
  },
  "deferred": [],
  "onattach": [
    [Function],
    [Function],
  ],
  "parse": [Function],
  "pattern": /\\^\\(\\?!\\\\\\.\\)\\(\\?!\\.\\*\\\\\\.\\\\\\.\\)\\(\\[A-Za-z0-9_'\\+\\\\-\\\\\\.\\]\\*\\)\\[A-Za-z0-9_\\+-\\]@\\(\\[A-Za-z0-9\\]\\[A-Za-z0-9\\\\-\\]\\*\\\\\\.\\)\\+\\[A-Za-z\\]\\{2,\\}\\$/,
  "run": [Function],
  "traits": Set {
    "ZodEmail",
    "$ZodEmail",
    "$ZodStringFormat",
    "$ZodCheckStringFormat",
    "$ZodCheck",
    "$ZodString",
    "$ZodType",
    "ZodStringFormat",
    "_ZodString",
    "ZodType",
  },
}
`;

exports[`Environment checks > Zod checks/refinements > Snapshot control 'ZodNumber' definition 1`] = `
{
  "bag": {
    "format": "safeint",
    "maximum": 9007199254740991,
    "minimum": -9007199254740991,
    "pattern": /\\^\\\\d\\+\\$/,
  },
  "constr": [Function],
  "def": {
    "checks": [
      {
        "$schema": "https://json-schema.org/draft-2020-12/schema",
        "maximum": 9007199254740991,
        "minimum": -9007199254740991,
        "type": "integer",
      },
    ],
    "type": "number",
  },
  "deferred": [],
  "parse": [Function],
  "pattern": /\\^\\\\d\\+\\$/,
  "run": [Function],
  "traits": Set {
    "ZodNumber",
    "$ZodNumber",
    "$ZodType",
    "ZodType",
  },
}
`;

exports[`Environment checks > Zod checks/refinements > Snapshot control 'ZodNumberFormat' definition 1`] = `
{
  "bag": {
    "format": "safeint",
    "maximum": 9007199254740991,
    "minimum": -9007199254740991,
    "pattern": /\\^\\\\d\\+\\$/,
  },
  "check": [Function],
  "constr": [Function],
  "def": {
    "abort": false,
    "check": "number_format",
    "format": "safeint",
    "type": "number",
  },
  "deferred": [],
  "onattach": [
    [Function],
  ],
  "parse": [Function],
  "pattern": /\\^\\\\d\\+\\$/,
  "run": [Function],
  "traits": Set {
    "ZodNumberFormat",
    "$ZodNumber",
    "$ZodCheckNumberFormat",
    "$ZodCheck",
    "$ZodType",
    "ZodNumber",
    "ZodType",
  },
}
`;

exports[`Environment checks > Zod checks/refinements > Snapshot control 'ZodNumberFormat' definition 2`] = `
{
  "bag": {
    "format": "int32",
    "maximum": 2147483647,
    "minimum": -2147483648,
    "pattern": /\\^\\\\d\\+\\$/,
  },
  "check": [Function],
  "constr": [Function],
  "def": {
    "abort": false,
    "check": "number_format",
    "format": "int32",
    "type": "number",
  },
  "deferred": [],
  "onattach": [
    [Function],
  ],
  "parse": [Function],
  "pattern": /\\^\\\\d\\+\\$/,
  "run": [Function],
  "traits": Set {
    "ZodNumberFormat",
    "$ZodNumber",
    "$ZodCheckNumberFormat",
    "$ZodCheck",
    "$ZodType",
    "ZodNumber",
    "ZodType",
  },
}
`;

exports[`Environment checks > Zod checks/refinements > Snapshot control 'ZodNumberFormat' definition 3`] = `
{
  "bag": {
    "format": "safeint",
    "maximum": 1000,
    "minimum": -9007199254740991,
    "pattern": /\\^\\\\d\\+\\$/,
  },
  "check": [Function],
  "constr": [Function],
  "def": {
    "abort": false,
    "check": "number_format",
    "checks": [
      $ZodCheckLessThan {},
    ],
    "format": "safeint",
    "type": "number",
  },
  "deferred": [],
  "onattach": [
    [Function],
  ],
  "parse": [Function],
  "pattern": /\\^\\\\d\\+\\$/,
  "run": [Function],
  "traits": Set {
    "ZodNumberFormat",
    "$ZodNumber",
    "$ZodCheckNumberFormat",
    "$ZodCheck",
    "$ZodType",
    "ZodNumber",
    "ZodType",
  },
}
`;

exports[`Environment checks > Zod checks/refinements > Snapshot control 'ZodString' definition 1`] = `
{
  "bag": {
    "format": "email",
    "pattern": /\\^\\(\\?!\\\\\\.\\)\\(\\?!\\.\\*\\\\\\.\\\\\\.\\)\\(\\[A-Za-z0-9_'\\+\\\\-\\\\\\.\\]\\*\\)\\[A-Za-z0-9_\\+-\\]@\\(\\[A-Za-z0-9\\]\\[A-Za-z0-9\\\\-\\]\\*\\\\\\.\\)\\+\\[A-Za-z\\]\\{2,\\}\\$/,
  },
  "constr": [Function],
  "def": {
    "checks": [
      {
        "$schema": "https://json-schema.org/draft-2020-12/schema",
        "format": "email",
        "pattern": "^(?!\\.)(?!.*\\.\\.)([A-Za-z0-9_'+\\-\\.]*)[A-Za-z0-9_+-]@([A-Za-z0-9][A-Za-z0-9\\-]*\\.)+[A-Za-z]{2,}$",
        "type": "string",
      },
    ],
    "type": "string",
  },
  "deferred": [],
  "parse": [Function],
  "pattern": /\\^\\(\\?!\\\\\\.\\)\\(\\?!\\.\\*\\\\\\.\\\\\\.\\)\\(\\[A-Za-z0-9_'\\+\\\\-\\\\\\.\\]\\*\\)\\[A-Za-z0-9_\\+-\\]@\\(\\[A-Za-z0-9\\]\\[A-Za-z0-9\\\\-\\]\\*\\\\\\.\\)\\+\\[A-Za-z\\]\\{2,\\}\\$/,
  "run": [Function],
  "traits": Set {
    "ZodString",
    "$ZodString",
    "$ZodType",
    "_ZodString",
    "ZodType",
  },
}
`;

exports[`Environment checks > Zod imperfections > circular object schema has no sign of getter in its shape 1`] = `
{
  "features": {
    "configurable": true,
    "enumerable": true,
    "value": {
      "$schema": "https://json-schema.org/draft-2020-12/schema",
      "items": {
        "properties": {
          "features": {
            "$ref": "#",
          },
          "name": {
            "type": "string",
          },
        },
        "required": [
          "name",
          "features",
        ],
        "type": "object",
      },
      "type": "array",
    },
    "writable": true,
  },
  "name": {
    "configurable": true,
    "enumerable": true,
    "value": {
      "$schema": "https://json-schema.org/draft-2020-12/schema",
      "type": "string",
    },
    "writable": true,
  },
}
`;

exports[`Environment checks > Zod imperfections > meta overrides, does not merge 1`] = `
{
  "title": "last",
}
`;

exports[`Environment checks > Zod new features > input examples of transformations 1`] = `
{
  "$schema": "https://json-schema.org/draft-2020-12/schema",
  "examples": [
    "test",
  ],
  "type": "string",
}
`;

exports[`Environment checks > Zod new features > output examples of transformations 1`] = `
{
  "$schema": "https://json-schema.org/draft-2020-12/schema",
  "examples": [
    4,
  ],
}
`;
